src/bsoncxx/CMakeLists.txt

```diff
@@ -78,39 +78,59 @@ set(LIBBSON_REQUIRED_ABI_VERSION 1.0)
 
 set(bsoncxx_pkg_dep "")
 
-# Attempt to find libbson by new package name (without lib).
-find_package(bson-${LIBBSON_REQUIRED_ABI_VERSION} ${LIBBSON_REQUIRED_VERSION} QUIET)
+if(BUNDLED_MONGOC AND (TARGET bson_shared OR TARGET bson_static))
+  # If these targets exist, then libbson has already been included as a project
+  # sub-directory
+  message ("found libbson targets declared in current build scope; version not checked")
+  message ("specify build option -D BUNDLED_MONGOC=OFF to disable")
 
-if(bson-${LIBBSON_REQUIRED_ABI_VERSION}_FOUND)
-  message ("found libbson version ${bson-${LIBBSON_REQUIRED_ABI_VERSION}_VERSION}")
   if(NOT BSONCXX_LINK_WITH_STATIC_MONGOC)
-    set(libbson_target mongo::bson_shared)
+    set(libbson_target bson_shared)
   else()
-    set(libbson_target mongo::bson_static)
+    set(libbson_target bson_static)
   endif()
+  # Ensures availability of bson.h via "#include <bson.h>"
+  get_property(LIBBSON_SRC_DIR TARGET bson_shared PROPERTY SOURCE_DIR)
+  set(EXTRA_LIBBSON_INCLUDE ${LIBBSON_SRC_DIR}/src/bson)
 
   if(BSONCXX_BUILD_STATIC)
     set(bsoncxx_pkg_dep "find_dependency(bson-1.0 REQUIRED)")
   endif()
 else()
-  # Require package of old libbson name (with lib).
-  if(NOT BSONCXX_LINK_WITH_STATIC_MONGOC)
-    find_package(libbson-${LIBBSON_REQUIRED_ABI_VERSION} ${LIBBSON_REQUIRED_VERSION} REQUIRED)
-    message ("found libbson version ${BSON_VERSION}")
-    set(libbson_target ${BSON_LIBRARIES})
-    set(libbson_include_directories ${BSON_INCLUDE_DIRS})
-    set(libbson_definitions ${BSON_DEFINITIONS})
+  # Attempt to find libbson by new package name (without lib).
+  find_package(bson-${LIBBSON_REQUIRED_ABI_VERSION} ${LIBBSON_REQUIRED_VERSION} QUIET)
+
+  if(bson-${LIBBSON_REQUIRED_ABI_VERSION}_FOUND)
+    message ("found libbson version ${bson-${LIBBSON_REQUIRED_ABI_VERSION}_VERSION}")
+    if(NOT BSONCXX_LINK_WITH_STATIC_MONGOC)
+      set(libbson_target mongo::bson_shared)
+    else()
+      set(libbson_target mongo::bson_static)
+    endif()
```

Comment:
Shouldn't bsoncxx_static link against bson_static, and bsoncxx_shared link against bson_shared, rather than both linking against the same kind of shared/static lib?

Reply:
Not necessarily.  It is possible to build a shared bsoncxx linked against bson_static.  That would encapsulate both bsoncxx and libbson together in a single shared object, where linking a shared bsoncxx to a shared libbson has a continuing requirement for both shared objects to be deployed.

Reply:
Does that mean that the bson_static files are always compiled as if they are to be integrated into a shared object, even if they are to be statically linked directly into a binary (ie with -fPIC rather than -fPIE)? Doesn't that add overhead that a consumer of the static archive would want to avoid? I seem to recall a small but measurable speed boost when we switched the server from -fPIC to -fPIE.

Reply:
In the C driver Evergreen build we only use the `-fPIE` option in the `hardened-compile` task.  The C++ driver build of the C driver uses `-fPIC` for all configurations (though the C++ driver Evergreen build does not have a `hardened-compile` task).  The C driver build also has an `ENABLE_PIC` option which defaults to `ON` and which adds `-fPIC` to the bson_static and mongoc_static builds.  In order to use `-fPIE` the user would have to include `-D ENABLE_PIC=OFF` and then add `-fPIE` to `CFLAGS`.  That's perhaps a long-winded way to say, "we always use `-fPIC`, but the user could override to use `-fPIE`".